File: Neuroflow/NeuralNetworks/Layer.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Neuroflow.NeuralNetworks
{
    public sealed class Layer
    {
        public Layer(int size)
        {
            Args.Requires(() => size, () => size > 0);

            Size = size;
            InputConnections = new LayerConnections(LayerConnections.ConnectionType.Input, this);
            OutputConnections = new LayerConnections(LayerConnections.ConnectionType.Output, this);
            Behaviors = new LayerBehavior.Collection();
            Descriptions = new LayerDescription.Collection();
        }

        public int Size { get; private set; }

        public LayerConnections InputConnections { get; private set; }

        public LayerConnections OutputConnections { get; private set; }

        public ICollection<LayerBehavior> Behaviors { get; private set; }

        public ICollection<LayerDescription> Descriptions { get; private set; }

        public bool HasRecurrentConnections
        {
            get
            {
                return InputConnections.GetConnectedLayers(FlowDirection.TwoWay | FlowDirection.OneWayToSource).Any() ||
                    OutputConnections.GetConnectedLayers(FlowDirection.TwoWay | FlowDirection.OneWayToSource).Any();
            }
        }

        public IEnumerable<Layer> GetInputLayers()
        {
            return InputConnections.GetConnectedLayers(FlowDirection.OneWay | FlowDirection.TwoWay)
                .Concat(OutputConnections.GetConnectedLayers(FlowDirection.TwoWay | FlowDirection.OneWayToSource));
        }

        public IEnumerable<Layer> GetOutputLayers()
        {
            return OutputConnections.GetConnectedLayers(FlowDirection.OneWay | FlowDirection.TwoWay)
                .Concat(InputConnections.GetConnectedLayers(FlowDirection.TwoWay | FlowDirection.OneWayToSource));
        }

        public Layer GetInputLayer(int connectionIndex)
        {
            var q = from item in GetInputLayers().Select((l, idx) => new { l, idx })
                    where item.idx == connectionIndex
                    select item.l;

            var layer = q.SingleOrDefault();

            if (layer == null) throw CreateLayerNotFoundException("Input", connectionIndex);

            return layer;
        }

        public Layer GetOutputLayer(int connectionIndex)
        {
            var q = from item in GetOutputLayers().Select((l, idx) => new { l, idx })
                    where item.idx == connectionIndex
                    select item.l;

            var layer = q.SingleOrDefault();

            if (layer == null) throw CreateLayerNotFoundException("Output", connectionIndex);

            return layer;
        }

        private Exception CreateLayerNotFoundException(string type, int connectionIndex)
        {
            return new ArgumentOutOfRangeException(type + " layer not found, connection index value " + connectionIndex + " was out of range.", "connectionIndex");
        }
    }
}

File: NeuroflowLib/layer.cpp
#include "stdafx.h"
#include "layer.h"

USING;

layer::layer(idx_t size) :
_inputConnections(layer_connections::connection_type::input),
_outputConnections(layer_connections::connection_type::output)
{
}

layer_description_coll& layer::descriptions()
{
    return _descriptions;
}

layer_behavior_coll& layer::behaviors()
{
    return _behaviors;
}

layer_connections& layer::input_connections()
{
    if (_inputConnections.must_init()) _inputConnections.init(shared_this<layer>());
    return _inputConnections;
}

layer_connections& layer::output_connections()
{
    if (_outputConnections.must_init()) _outputConnections.init(shared_this<layer>());
    return _outputConnections;
}

bool layer::has_recurrent_connections() const
{
    return _inputConnections.connected_layers(flow_direction::one_way_to_source | flow_direction::two_way) | any()
        || _outputConnections.connected_layers(flow_direction::one_way_to_source | flow_direction::two_way) | any();
}

layers_t layer::input_layers() const
{
    return _inputConnections.connected_layers(flow_direction::one_way | flow_direction::two_way)
        | concat(_outputConnections.connected_layers(flow_direction::two_way | flow_direction::one_way_to_source));
}

layers_t layer::output_layers() const
{
    return _outputConnections.connected_layers(flow_direction::one_way | flow_direction::two_way)
        | concat(_inputConnections.connected_layers(flow_direction::two_way | flow_direction::one_way_to_source));
}

layer_ptr layer::get_input_layer(idx_t connectionIndex) const
{
    auto result = input_layers()
        | row_num()
        | where([=](row_numbered<layer_ptr>& obj) { return obj.row_num() == connectionIndex; })
        | select([](row_numbered<layer_ptr>& obj) { return obj.value(); })
        | first_or_default();

    if (!result) throw_logic_error("Input layer not found, connection index value " + to_string(connectionIndex) + " was out of range.");
    return result;
}

layer_ptr layer::get_output_layer(idx_t connectionIndex) const
{
    auto result = output_layers()
        | row_num()
        | where([=](row_numbered<layer_ptr>& obj) { return obj.row_num() == connectionIndex; })
        | select([](row_numbered<layer_ptr>& obj) { return obj.value(); })
        | first_or_default();

    if (!result) throw_logic_error("Output layer not found, connection index value " + to_string(connectionIndex) + " was out of range.");
    return result;
}